libdnf/utils/exception.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright (C) 2021 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf/utils/exception.hpp"
+
+#include <charconv>
+#include <system_error>
+
+
+namespace libdnf {
+
+
+const char * Exception::what() const noexcept {
+    try {
+        str_what = std::string(get_domain_name()) + "::" + get_name() + ": " + get_description();
+
+        // related data, eg file name
+        const char * what = runtime_error::what();
+        if (what && what[0] != '\0') {
+            str_what += ": ";
+            str_what += what;
+        }
+
+        return str_what.c_str();
+    } catch (...) {
+        return runtime_error::what();
+    }
+}
+
+const char * SystemError::get_name() const noexcept {
+    std::to_chars(name + 6, name + sizeof(description) - 1, get_error_code());
+    return name;
+}
+
+const char * SystemError::get_description() const noexcept {
+    try {
+        description = std::system_category().default_error_condition(error_code).message();
+    } catch (...) {
```

Comment:
Do we know what we are catching here? I don't like `catch (...)` and would only use it in as few places as possible, which would seem to be in `Exception::what()`.

Reply:
The `message()` method returns an explanatory message for the stored error code and error category. 
However, according to the standard, it may throw implementation-defined exceptions. Therefore, all exceptions are caught with `catch(...)` and an explanatory message "Unknown error" is returned.